### DISCUSSION TITLE
Flash von Docker oder ohne build?

### DISCUSSION THREAD

**Discussion by vogler** [Freetz-NG/freetz-ng#622] [Allgemein] [answered]
Auf meinem M1 MBA hat der build via Docker/qemu leider ~6h gedauert. Kann ich das image flashen ohne das Ganze nochmal machen zu müssen?

Über die FritzBox selbst ging nicht:
![image](https://user-images.githubusercontent.com/493741/193017121-0d1b2d98-5891-4839-865d-b001cf1628d0.png)

Dann vom Container:
```console
builduser@e57a64a95795:~$ ./tools/push_firmware

 * Analyzing 'images/7430_07.29.all_freetz-ng-20254M-342950c57_20220928-181702.image' ...

 * Warning: It seems your network is not able to reach
   192.168.178.1 directly. One(!) of these commands could help to fix:
   $ sudo ifconfig eth0:0 192.168.178.2 up
   $ sudo ifconfig ip6tnl0:0 192.168.178. up
   $ sudo ifconfig tunl0:0 192.168.178. up
   Proceed anyway? ([y]/n)
 * Product: Fritz_Box_HW218 (FRITZ!Box 7430)
 * Using command: ftp
 * Target host: 192.168.178.1
 * Outgoing IP: unknown
 * Flash mode: ram-boot
 * Allowed memory size: 128 MB
 * Designated linux_fs_start: <other>

 !!! WARNING !!! WARNING !!! WARNING !!! WARNING !!! WARNING !!!
 !!!  THERE IS NO WARRANTY AT ALL !!! USE AT YOUR OWN RISK   !!!

 * Are you sure, that you want to flash this file to the device?
   images/7430_07.29.all_freetz-ng-20254M-342950c57_20220928-181702.image
   Proceed? (y/[n]) y

 * You should now reboot your box (192.168.178.1). Waiting for shut down.
   Switch off, if reboot is not detected because it happens too quickly.
   Some newer bootloader versions allow to flash on power-cycle only.

 * MAPSTART=0x80000000
 * FULLSIZE=0x08000000	 (128 MB)
 * MAPLIMIT=0x88000000
 * FILESIZE=0x01da5d00	 (~29 MB)
 * ALIBYTES=0x00000000	 (~0 kB)
 * FREESIZE=0x0625a300	 (~98 MB)
 * MTDSTART=0x8625a300

 * No reply from box, assuming switch-off or restart. Trying to re-detect box.
   Waiting .................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................^C
aborted
builduser@e57a64a95795:~$ ping 192.168.178.1
PING 192.168.178.1 (192.168.178.1) 56(84) bytes of data.
^C
--- 192.168.178.1 ping statistics ---
3 packets transmitted, 0 received, 100% packet loss, time 2085ms

builduser@e57a64a95795:~$ ifconfig
eth0: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet 172.17.0.2  netmask 255.255.0.0  broadcast 172.17.255.255
        ether 02:42:ac:11:00:02  txqueuelen 0  (Ethernet)
        RX packets 271016  bytes 374799828 (374.7 MB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 122747  bytes 38302842 (38.3 MB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

lo: flags=73<UP,LOOPBACK,RUNNING>  mtu 65536
        inet 127.0.0.1  netmask 255.0.0.0
        loop  txqueuelen 1000  (Local Loopback)
        RX packets 0  bytes 0 (0.0 B)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 0  bytes 0 (0.0 B)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0
```

Vom host (192.168.178.1 erreichbar):
```console
❯ ./tools/push_firmware

 * Analyzing 'images/7430_07.29.all_freetz-ng-20254M-342950c57_20220928-181702.image' ...
Can not recognize image, please use '-m<s|r|d|u|f>'.

❯ ./tools/push_firmware -mr

Error: file is not a valid ram-boot image to be written to the device.
```
Brauche mehr als nur die .image Datei oder hat das Script ein Problem mit macOS?

Nun bin ich verwirrt 😄

**Comment by nixmomo** [comment]
Der Container muss im Bridged Mode laufen sonst erreicht der die Fritte nicht

**Reply by vogler** [reply]
Ok, probiere ich dann mal, anscheinend geht das auch bei laufenden Containern: https://docs.docker.com/network/bridge/
Ist es tatsächlich der Fall dass man mehr als die .image Datei braucht um erfolgreich zu flashen?

**Reply by pfichtner** [reply]
Bridge Modus sollte eigentlich default sein: 
"bridge: The default network driver. If you don’t specify a driver, this is the type of network you are creating. Bridge networks are usually used when your applications run in standalone containers that need to communicate. See [bridge networks](https://docs.docker.com/network/bridge/)."
Quelle: https://docs.docker.com/network/#network-drivers

**Reply by pfichtner** [reply]
Ich vermute eher, dass zwischen dem Docker Container und dem Rechner/der NIC nochmal einiges an VIrtualisierungsschichten liegt die hier die Probleme verursachen

**Reply by vogler** [reply]
Ja, `bridge` ist wohl default und hilft nicht. Habe es auch mit `docker network create/connect` versucht. `--network=host` würde evtl. helfen, aber das gibt es nicht bei macOS Docker.
`gateway.docker.internal` ist zwar erreichbar aber ging auch nicht:
```
$ ping gateway.docker.internal
PING gateway.docker.internal (192.168.65.2) 56(84) bytes of data.
64 bytes from 192.168.65.2: icmp_seq=1 ttl=37 time=0.453 ms
$ ./tools/push_firmware -ip gateway.docker.internal

 * Analyzing 'images/7430_07.29.all_freetz-ng-20254M-342950c57_20220928-181702.image' ...

 * Warning: It seems your network is not able to reach
   gateway.docker.internal directly. One(!) of these commands could help to fix:
   $ sudo ifconfig eth0:0 gateway.docker.2 up
   $ sudo ifconfig eth1:0 gateway.docker.2 up
   $ sudo ifconfig ip6tnl0:0 gateway.docker. up
   $ sudo ifconfig tunl0:0 gateway.docker. up
   Proceed anyway? ([y]/n)
 * Product: Fritz_Box_HW218 (FRITZ!Box 7430)
 * Using command: ftp
 * Target host: gateway.docker.internal
 * Outgoing IP: unknown
 * Flash mode: ram-boot
 * Allowed memory size: 128 MB
 * Designated linux_fs_start: <other>

 !!! WARNING !!! WARNING !!! WARNING !!! WARNING !!! WARNING !!!
 !!!  THERE IS NO WARRANTY AT ALL !!! USE AT YOUR OWN RISK   !!!

 * Are you sure, that you want to flash this file to the device?
   images/7430_07.29.all_freetz-ng-20254M-342950c57_20220928-181702.image
   Proceed? (y/[n]) y

 * You should now reboot your box (gateway.docker.internal). Waiting for shut down.
   Switch off, if reboot is not detected because it happens too quickly.
   Some newer bootloader versions allow to flash on power-cycle only.
   .......................................

**Reply by vogler** [reply]
Docker von macOS aus gebe ich auf. VirtualBox VM auf Windows neu gebaut und er erkennt die FritzBox.

Muss ich bei der 7430 irgendetwas beachten? Ich habe für ein paar Sekunden den Stecker gezogen, aber es passiert nichts:
![image](https://user-images.githubusercontent.com/493741/194095482-95b421a7-ed43-4b8d-8a90-b533f992ef65.png)

**Reply by vogler** [reply]
Ich bin nun [dieser Anleitung](https://web.archive.org/web/20190529205946/https://www.ip-forum.eu/howto-fritz-box-firmware-images-auch-unsignierte-ueber-den-bootloader-installieren-577) von 2018 gefolgt und konnte es unter Windows flashen.

In VM noch `./tools/image2inmemory` und das Ergebnis kopiert.
Dann in Windows Powershell:
```console
$ git clone https://github.com/PeterPawn/YourFritz.git
$ cd YourFritz\eva_tools
$ .\EVA-Discover.ps1 -maxWait 120 -Debug -Verbose
DEBUG: Using local IPv4 address 192.168.178.5 ...
VERBOSE: Sending discovery packet (1) ...
# Stecker raus/rein
DEBUG: Received UDP packet from 192.168.178.1:5035 ...
VERBOSE: Found EVA loader at 192.168.178.1 ...
VERBOSE: Trying to connect to the FTP port to hold up the device in bootloader ...
EVA_IP=192.168.178.1
True
$ ./EVA-FTP-Client.ps1 -Verbose -Debug -ScriptBlock { BootDeviceFromImage C:\Users\Ralf\Downloads\7430_07.29.all_freetz-ng-20290-c405257da_20221005-144441.image.inmemory }
DEBUG: Response:
220 ADAM2 FTP Server ready
...
226 Transfer complete
```

**Comment by Prisrak1** [comment]
ist doch erreichbar ->"Vom host (192.168.178.1 erreichbar):" und ins ram kommt doch was entpacktes rein, so etwas wie "*.image.in-memory"

somit ist was falsch in der Befehlseingabe. (da gibts den Unterschied zwischen "./tools/push_firmware" etwas differenzierte Einstellungen und "make push-firmware" -> gerade gebautes Image)

war mal so frei: 
Error: file is not a valid ram-boot image to be written to the device.
Datei ist kein gültiges Ram-Boot-Image, das auf das Gerät geschrieben werden kann.

**Reply by vogler** [reply]
Verstehe nicht was du meinst.
Von Docker aus war die FritzBox nicht erreichbar weil keine bridge. Image wurde als `Flash mode: ram-boot` erkannt.
Von macOS aus ist die FritzBox zwar erreichbar, aber das Script beschwert sich über das Image. Bei den anderen `-m` Optionen ebenso.

**Comment by vogler** [comment, marked as answer]
Also bei mir ging weder von Docker unter macOS noch von VirtualBox unter Windows.
Mit Windows Powershell die in der VM erzeugte `.image.inmemory` flashen hat dann funktioniert.
Siehe https://github.com/Freetz-NG/freetz-ng/discussions/622#discussioncomment-3807253
